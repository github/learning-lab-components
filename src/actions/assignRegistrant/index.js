const has = require('has')

module.exports = async (context, opts) => {
  let number
  if (has(opts, 'issue')) {
    if (typeof opts.issue === 'string') {
      const { owner, repo } = context.repo()
      const issues = await context.github.search.issues({
        q: `in:title ${opts.issue} repo:${owner}/${repo}`
      })
      const issue = issues.data.items.find(i => i.title === opts.issue)
      if (issue) number = issue.number
    } else {
      number = opts.issue
    }

    if (!number) throw new Error('Issue could not be found')
  } else {
    number = context.issue().number
  }

  return context.github.request(context.repo({
    method: 'POST',
    url: '/repos/:owner/:repo/issues/:number/assignees',
    headers: { accept: 'application/vnd.github.symmetra-preview+json' },
    number,
    assignees: [context.user.login]
  }))
}