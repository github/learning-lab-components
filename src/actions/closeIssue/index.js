const has = require('has')

module.exports = async (context, opts) => {
  let number
  if (has(opts, 'issue')) {
    if (typeof opts.issue === 'string') {
      const { owner, repo } = context.repo()
      const issues = await context.github.search.issues({
        q: `is:issue in:title ${opts.issue} repo:${owner}/${repo}`
      })
      const issue = issues.data.items.find(i => i.title === opts.issue)
      if (issue) number = issue.number
    } else {
      number = opts.issue
    }

    if (!number) throw new Error('Issue could not be found')
  } else {
    number = context.issue().number
  }

  return context.github.issues.update(context.repo({
    number,
    state: 'closed'
  }))
}
