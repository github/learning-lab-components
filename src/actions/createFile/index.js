module.exports = async (context, opts) => {
  // Fill in template variables
  const content = await context.fromFile(opts.filename, opts.data)

  // Convert to base64 for GitHub
  const blob64 = Buffer.from(content).toString('base64')

  const path = opts.new_name || opts.filename
  const message = opts.message || `Creating new file ${path}`

  // Prepare params
  const params = context.repo({
    path,
    message,
    content: blob64
  })

  // If no branch is provided, GitHub will use the default_branch
  if (opts.branch) params.branch = opts.branch

  return context.github.repos.createFile(params)
}
