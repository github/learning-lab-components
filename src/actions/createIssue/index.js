const has = require('has')

module.exports = async (context, opts) => {
  const issue = await context.github.issues.create(context.repo({
    title: opts.title,
    body: await context.fromFile(opts.body, opts.data)
  }))

  if (has(opts, 'comments')) {
    for (const comment of opts.comments) {
      await context.github.issues.createComment(context.repo({
        number: issue.data.number,
        body: await context.fromFile(comment, opts.data)
      }))
    }
  }

  return issue
}
