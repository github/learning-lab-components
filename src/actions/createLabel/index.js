module.exports = async (context, opts) => {
  return context.github.issues.createLabel(context.repo({
    name: opts.name,
    color: opts.color,
    description: opts.description
  }))
}
