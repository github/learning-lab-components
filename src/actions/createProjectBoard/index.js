const has = require('has')

module.exports = async (context, opts) => {
  // Create a new project board
  const newProjectBoard = await context.github.projects.createRepoProject(context.repo({
    name: opts.name,
    body: opts.description
  }))

  // Bootstrap it with the appropriate columns
  if (has(opts, 'columns') && Array.isArray(opts.columns)) {
    for (const column of opts.columns) {
      await context.github.projects.createProjectColumn({
        project_id: newProjectBoard.data.id,
        name: column
      })
    }
  }

  // Return the project board itself
  return newProjectBoard
}
