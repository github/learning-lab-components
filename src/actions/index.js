const fs = require('fs')

const actionNames =
  fs.readdirSync(__dirname, { withFileTypes: true })
    .filter(ent => ent.isDirectory())
    .map(ent => ent.name)
    .sort()

const actions = {}

for (const actionName of actionNames) {
  actions[actionName] = require(`./${actionName}`)
  actions[actionName].schema = require(`./${actionName}/schema`)
}

module.exports = actions
