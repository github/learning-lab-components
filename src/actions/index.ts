import fs from 'fs';

import * as types from '../types';

const actionNames =
  fs.readdirSync(__dirname, { withFileTypes: true })
    .filter(ent => ent.isDirectory())
    .map(ent => ent.name)
    .sort()

let actions: types.IAction = {};

for (const actionName of actionNames) {
  actions[actionName] = require(`./${actionName}`)
  actions[actionName].schema = require(`./${actionName}/schema`);
}

export default actions;
