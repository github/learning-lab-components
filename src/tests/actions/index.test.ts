import path from 'path';
import fs from 'fs';
import isPlainObject from 'lodash.isplainobject';

import actions from '../../actions';

const actionsDir = path.join(__dirname, '../../actions')

const actionNames =
  fs.readdirSync(actionsDir, { withFileTypes: true })
    .filter(ent => ent.isDirectory())
    .map(ent => ent.name)
    .sort()

describe('actions', () => {
  it('module has properties for all expected actions', async () => {
    const actionKeys = Object.keys(actions).sort()
    expect(actionKeys).toEqual(actionNames)
    expect(actionKeys).toMatchSnapshot()
  })

  it('README mentions all expected actions', async () => {
    const actionsReadme = fs.readFileSync(path.join(actionsDir, 'README.md'), { encoding: 'utf8' })

    const expectedContents = `## Available actions

<!-- START_ACTIONS_LIST -->
${actionNames.map(name => `- [${name}](./${name})`).join('\n')}

<!-- END_ACTIONS_LIST -->
`
    // Replacing new line before comparing.
    expect(actionsReadme.replace(/(\r\n|\n|\r)/gm,"")).toContain(expectedContents.replace(/(\r\n|\n|\r)/gm,""))
  })

  describe.each(actionNames)('the action "%s"', (actionName) => {
    const action = actions[actionName]

    it('has the minimum set of required files', async () => {
      const actionDirFiles =
        fs.readdirSync(path.join(actionsDir, actionName), { withFileTypes: true })
          .filter(ent => ent.isFile())
          .map(ent => ent.name)

      expect(actionDirFiles).toContain('schema.js')
      expect(actionDirFiles).toContain('index.js')
      expect(actionDirFiles).toContain('index.test.js')
      expect(actionDirFiles).toContain('README.md')
    })

    it('is a function expecting 2-3 arguments', () => {
      expect(typeof action).toBe('function')
      expect(action.length).toBeGreaterThanOrEqual(2)
      expect(action.length).toBeLessThanOrEqual(3)
    })

    it('has a "schema" property', () => {
      expect(action).toHaveProperty('schema')
      expect(action.schema).toBeTruthy()
    })

    describe('has a schema which', () => {
      // Get a plain object rendering of the schema to avoid relying on Joi internals
      const schema = action.schema.describe()

      it('has a description', () => {
        expect(typeof schema.description).toBe('string')
        expect(schema.description.length).toBeGreaterThan(0)
      })

      it('has at least one example', () => {
        expect(Array.isArray(schema.examples)).toBe(true)
        expect(schema.examples.length).toBeGreaterThan(0)
      })

      it('has examples which all include a context if there is more than one example', () => {
        const hasMoreThanOneExample = schema.examples.length > 1

        for (const example of schema.examples) {
          expect(Array.isArray(example)).toBe(true)
          expect(example.length).toBeGreaterThanOrEqual(1)
          expect(example.length).toBeLessThanOrEqual(2)

          const [value, options] = example
          expect(isPlainObject(value)).toBe(true)

          if (hasMoreThanOneExample) {
            expect(isPlainObject(options)).toBe(true)
            expect(options).toHaveProperty('context')
            expect(typeof options.context).toBe('string')
            expect(options.context.length).toBeGreaterThan(0)
          }
        }
      })
    })
  })
})
