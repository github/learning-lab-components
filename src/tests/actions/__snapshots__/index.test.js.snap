// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions module has properties for all expected actions 1`] = `
Array [
  "assignRegistrant",
  "closeIssue",
  "createFile",
  "createIssue",
  "createLabel",
  "createProjectBoard",
  "createPullRequest",
  "createPullRequestComment",
  "createReview",
  "createStatus",
  "deleteBranch",
  "findInTree",
  "gate",
  "getFileContents",
  "getIssue",
  "getPullRequest",
  "getTree",
  "htmlContainsTag",
  "mergeBranch",
  "mergePullRequest",
  "octokit",
  "removeBranchProtection",
  "requestReviewFromRegistrant",
  "respond",
  "updateBranchProtection",
]
`;
